hmac_authentication/hmacauth.py

```diff
@@ -1,57 +1,65 @@
+
+
+class AuthenticationResultCodes(enum.Enum):
+    '''Defines the result codes used in AuthenticationResult.'''
 
+    # The incoming result did not have a signature header
     NO_SIGNATURE = 1
+
+    # The signature header was not parseable
     INVALID_FORMAT = 2
+
+    # The signature header specified an unsupported algorithm
     UNSUPPORTED_ALGORITHM = 3
+
+    # The signature from the request header matched the locally-computed
+    # signature
     MATCH = 4
+
+    # The signature from the request header did not match the locally-computed
+    # signature
     MISMATCH = 5
 
-    RESULT_CODE_STRINGS = [
-        'NO_SIGNATURE',
-        'INVALID_FORMAT',
-        'UNSUPPORTED_ALGORITHM',
-        'MATCH',
-        'MISMATCH',
-    ]
-
-    @staticmethod
-    def result_code_to_string(code):
-        '''Maps one of the constants representing validate_request() return
-        codes to a string representation from RESULT_CODE_STRINGS.
-        '''
-        index = code - 1
-        if index >= 0 and index < len(HmacAuth.RESULT_CODE_STRINGS):
-            return HmacAuth.RESULT_CODE_STRINGS[index]
-
-    @staticmethod
-    def header_name_to_wsgi(header_name):
-        wsgi_name = header_name.upper().replace('-', '_')
-        if wsgi_name not in ['CONTENT_TYPE', 'CONTENT_LENGTH']:
-            wsgi_name = 'HTTP_' + wsgi_name
-        return wsgi_name
-
-    def __init__(self, digest_name, secret_key, signature_header, headers):
-        if not digest_name in hashlib.algorithms_available:
-            raise exceptions.Error(
-                'HMAC authentication digest is not supported: ' + digest_name)
-        self._digest = getattr(hashlib, digest_name)
-        self._digest_name = digest_name
+
+def header_name_to_wsgi(header_name):
+    wsgi_name = header_name.upper().replace('-', '_')
+    if wsgi_name not in ['CONTENT_TYPE', 'CONTENT_LENGTH']:
+        wsgi_name = 'HTTP_' + wsgi_name
+    return wsgi_name
+
+
+def _compare_signatures(header, computed):
+    if hmac.compare_digest(header.encode('utf8'), computed.encode('utf8')):
+        return AuthenticationResultCodes.MATCH
+    return AuthenticationResultCodes.MISMATCH
+
+
+class HmacAuth(object):
+    '''HmacAuth signs outbound requests and authenticates inbound requests.
+
+    Note that the method parameters called "req" or "request" correspond to
+    the WSGI "environ" interface define in PEP 333:
+      https://www.python.org/dev/peps/pep-0333/
+    '''
+
+    def __init__(self, digest, secret_key, signature_header, headers):
+        self._digest = digest
         self._secret_key = secret_key
-        self._signature_header = HmacAuth.header_name_to_wsgi(signature_header)
-        self._headers = [HmacAuth.header_name_to_wsgi(h) for h in headers]
+        self._signature_header = header_name_to_wsgi(signature_header)
+        self._headers = [header_name_to_wsgi(h) for h in headers]
 
     # NOTE(mbland): Best I can tell, either the WSGI layer or every sane HTTP
```

Comment:
Yep, it looks like wsgi-compliant apps are supposed to join repeated headers by commas: http://stackoverflow.com/questions/1801124/how-does-wsgi-handle-multiple-request-headers-with-the-same-name.